simple/rdm_shared_av.c

```diff
@@ -0,0 +1,237 @@
+		return ret;
+
+	if (opts.dst_addr && !parent) {
+		/* child waits until parent is done creating AV */
+		ret = read(pair[0], &ready, sizeof(int));
+		if (ret < 0)
+			FT_PRINTERR("read", errno);
+
+		/* child needs to open AV in read only mode */
+		av_attr.flags = FI_READ;
+	}
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	if (opts.dst_addr && parent) {
+		/* parent lets child know that AV is created */
+		ret = write(pair[1], &ready, sizeof(int));
+		if (ret < 0)
+			FT_PRINTERR("write", errno);
+	}
+
+	ret = ft_init_ep();
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static int send_recv()
+{
+	struct fi_cq_entry comp;
+	int ret;
+
+	if (opts.dst_addr) {
+		fprintf(stdout, "Sending message...\n");
+		snprintf(buf, buf_size, "Hello from Child Client!");
+
+		ret = fi_send(ep, buf, strlen(buf),
+				fi_mr_desc(mr), remote_fi_addr, &tx_ctx);
+		if (ret) {
+			FT_PRINTERR("fi_send", ret);
+			return ret;
+		}
+
+		do {
+			ret = fi_cq_read(txcq, &comp, 1);
+			if (ret < 0 && ret != -FI_EAGAIN) {
+				FT_PRINTERR("fi_cq_read", ret);
+				return ret;
+			}
+		} while (ret == -FI_EAGAIN);
+
+		fprintf(stdout, "Send completion received\n");
+	} else {
+		fprintf(stdout, "Waiting for message from client...\n");
+		do {
+			ret = fi_cq_read(rxcq, &comp, 1);
+			if (ret < 0 && ret != -FI_EAGAIN) {
+				FT_PRINTERR("fi_cq_read", ret);
+				return ret;
+			}
+		} while (ret == -FI_EAGAIN);
+
+		fprintf(stdout, "Received data from client: %s\n", (char *)buf);
+	}
+
+	return 0;
+}
+
+static int run(void)
+{
+	int ret;
+
+	ret = init_fabric();
+	if (ret)
+		return ret;
+
+	if (opts.dst_addr) {
+		if (parent) {
+			/* parent inits AV and lets child proceed,
+			 * and itself returns without sending a message */
+			ret = ft_init_av();
+			if (ret)
+				return ret;
+			ret = write(pair[1], &ready, sizeof(int));
+			if (ret < 0)
+				FT_PRINTERR("write", errno);
+
+			return 0;
+		} else {
+			/* client: child waits for parent to complete av_insert */
+			ret = read(pair[0], &ready, sizeof(int));
+			if (ret < 0)
+				FT_PRINTERR("read", errno);
+
+			remote_fi_addr = ((fi_addr_t *)av_attr.map_addr)[0];
+		}
+	} else {
+		ret = ft_init_av();
+		if (ret)
+			return ret;
+	}
+
+	return send_recv();
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+	pid_t child_pid = 0;
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "h" ADDR_OPTS INFO_OPTS)) != -1) {
+		switch (op) {
+		default:
+			ft_parse_addr_opts(op, optarg, &opts);
+			ft_parseinfo(op, optarg, hints);
+			ft_parsecsopts(op, optarg, &opts);
+			break;
+		case '?':
+		case 'h':
+			ft_usage(argv[0], "A shared AV client-sever example.");
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		opts.dst_addr = argv[optind];
+
+	if (opts.dst_addr) {
+		ret = socketpair(AF_LOCAL, SOCK_STREAM, 0, pair);
+		if (ret)
+			FT_PRINTERR("socketpair", errno);
+		child_pid = fork();
```

Comment:
Can this please be done without invoking fork?  fork causes bad things to happen with the providers.